s2cloudless/cloud_detector.py

```diff
@@ -0,0 +1,162 @@
+                      threshold value are masked as cloudy pixels. Default is 0.4.
+    :type threshold: float
+    :param all_bands: Flag specifying that input images will consists of all 13 Sentinel-2 bands.
+    :type all_bands: bool
+    :param average_over: Size of the disk in pixels for performing convolution (averaging probability
+                         over pixels). Value 0 means do not perform this post-processing step.
+                         Default is 1.
+    :type average_over: int
+    :param dilation_size: Size of the disk in pixels for performing dilation. Value 0 means do not perform
+                          this post-processing step. Default is 1.
+    :type dilation_size: int
+    :param model_filename: Location of the serialised model. If None the default model provided with the
+                           package is loaded.
+    :type model_filename: str or None
+    """
+
+    def __init__(self, threshold=0.4, all_bands=False, average_over=1, dilation_size=1, model_filename=None):
+
+        self.threshold = threshold
+        self.all_bands = all_bands
+        self.average_over = average_over
+        self.dilation_size = dilation_size
+
+        if model_filename is None:
+            package_dir = os.path.dirname(__file__)
+            model_filename = os.path.join(package_dir, 'models', MODEL_FILENAME)
+        self.model_filename = model_filename
+
+        self._classifier = None
+
+        if average_over > 0:
+            self.conv_filter = disk(average_over) / np.sum(disk(average_over))
+
+        if dilation_size > 0:
+            self.dilation_filter = disk(dilation_size)
+
+    @property
+    def classifier(self):
+        """
+        Provides a classifier object. It also loads it if it hasn't been loaded yet. This way the classifier is loaded
+        only when it is actually required.
+        """
+        if self._classifier is None:
+            self._classifier = PixelClassifier(Booster(model_file=self.model_filename))
+
+        return self._classifier
+
+    def get_cloud_probability_maps(self, data, **kwargs):
+        """
+        Runs the cloud detection on the input images (dimension n_images x n x m x 10
+        or n_images x n x m x 13) and returns an array of cloud probability maps (dimension
+        n_images x n x m). Pixel values close to 0 indicate clear-sky-like pixels, while
+        values close to 1 indicate pixels covered with clouds.
+
+        :param data: input Sentinel-2 image obtained with Sentinel-Hub's WMS/WCS request
+                  (see https://github.com/sentinel-hub/sentinelhub-py)
+        :type data: numpy array (shape n_images x n x m x 10 or n x m x 13)
+        :param kwargs: Any keyword arguments that will be passed to the classifier's prediction method
+        :return: cloud probability map
+        :rtype: numpy array (shape n_images x n x m)
+        """
+        is_single_temporal = data.ndim == 3
```

Comment:
So that it can work also on a single image and not only a stack of images. It is something that I'm using in the example notebook and find convenient.

Reply:
Shouldn't that be somewhat resolved in `_download_bands_and_valid_data_mask`? So that data is always a temporal array, in this case of single element?

_Whenever I see the same check being done over and over again, I am sure it should be "solved" at some other place._

Another consideration: how did it work so far? AFAIK the code so far worked even for single image...

Reply:
I see now that this is also algorithmically wrong since `get_mask_from_prob` expects a temporal component. Will fix.

Reply:
No, actually it is correct. I double-checked now and even added a unit test that proves that.

To mediate these few lines of duplication I would either have to make about 3 new methods or a decorator. Either option seems like an overkill. So I would just leave it for now.